
<img src = img/title.png >

# Countdown Numbers Game

***

## Overview and explanation
***

### What is it?


Countdown is a British game show of letters, words and numbers. In the numbers part of the game the players must try to combine six selected numbers using the four elementary mathematical operations to get as close as possilbe to the randomly generated three-digit target number.

### The rules

A contestant starts the numbers game by first picking how many small or large numbers they would like to make up the six numbers. There are 20 small numbers available ranging between 1 and 10 twice and 4 large numbers available, being 25, 50, 75 and 100. All large numbers must be different, so at most 4 large numbers can be chosen. The target number is randomly generated between the range of 101 to 999, the contestants must use basic arithemtic (+, -, × , ÷)on the generated numbers to get as close as possible to the target number, fractions or negative integers are not allowed in any stage of the calculation. You can use as many of the six selected numbers as you want, but can only use each number once.

## Discussion of the complexity
***

The implementation of a Countdown numbers game solver is a problematic and complex challenge. The first thing to discuss is the size of the problem and how many different possible solutions could exist. 

In [57]:
# Number of permutations of the six generated playing numbers.
ops = [1, 2, 5, 10, 75, 100]
nPermutations = 0
for q in it.permutations(ops, 6):
    nPermutations = nPermutations+1
print(nPermutations) 

720


In [58]:
# Number of combinations of the five operators with replacement.
ops = ['+', '*', '-', '/']
oCombinations = 0
for q in it.product(ops, repeat=5):
    oCombinations = oCombinations+1
print(oCombinations) 

1024


In [59]:
# Number of combinations of RPN
RPN = 42

In [60]:
# Total number of possible solutions to a single generated playing numbers set.
totalCombinations = nPermutations * oCombination * RPN
totalCombinations

30965760

Out of the six selected numbers, there are 720 permutations of playing numbers. 1024 different combinations to the five operators that can be used within the calculation. And 42 different ways to format the calculation into a Reverse Polish Notation. In total there are **30,965,760** different ways to write the calculation for the numbers game. However this total only considers one generation set of playing numbers, for every possible set of playing numbers the total is far greater.

In [61]:
# From the Python Standard Library.
import itertools as it

# Find the number of possible generated playing numbers for any game
playingSet = [1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9,10,10,25,50,75,100]
generatedSet = 0
for c in it.combinations(playingSet, 6):
    generatedSet = generatedSet+1
    
print(generatedSet)

134596


In [62]:
# Total number of possible solutions to all possible playing numbers.
generatedSet * operatorCombo * RPN

5788704768

Out of all the possible generated sets of playing numbers there are a total of **5,788,704,768** possible solutions. However this can be trimmed significantly as negatives or fractions are not allowed in any stage of the calculation.

## Pyton function to solve a Countdown numbers game
***

The function should take a list of six numbers and a target
number and return a method to calculate the target from the numbers, if it exists.

## Simulate a game

In [3]:
# For random numbers and samples.
import random

def new_numbers_game(no_large=None):
    """ Returns six numbers and a target number representing a Countdown numbers game.
    """
    # If no_large is None, randomly pick value between 0 and 4 inclusive.
    if no_large is None:
        # Randomly set the value.
        no_large = random.randrange(0, 5)
        
    # Select random large numbers.
    large_rand = random.sample([25, 50, 75, 100], no_large)
    # Select random small numbers.
    small_rand = random.sample(list(range(1, 11)) * 2, 6 - no_large)
    # The playing numbers.
    play_num = large_rand + small_rand
    
        
    # Select a target number.
    target = random.randrange(101, 1000)
    
    # Return the game.
    return play_num, target

In [12]:
# Random numbers game.
play_num, target = new_numbers_game()
play_num, target

([100, 4, 2, 2, 3, 8], 714)

## The solution

In [26]:
# Give all 2-partitions of a list 
# where each sublist has one element.
def patterns(numbers, operators):
    # Check if there is no way to partition further.
    if len(numbers) == 1:
        yield numbers
    # Loop through all the ways to partition L into two non-empty sublists
    for i in range(1, len(numbers)):
        # Slice the list using i.
        for left, right in it.product(patterns(numbers[:i], operators[1:i]), patterns(numbers[i:], operators[i:])):
            # Yield the next operator applied to the sublists.
            yield [*left, *right, operators[0]]

In [27]:
def eval_rpn(rpn):
    # A stack.
    stack = []
    # Loop through rpn an item at a time.
    for i in rpn:
        # Check if it's a number.
        if isinstance(i, int):
            # Append to the stack.
            stack = stack + [i]
        else:
            # Pop from stack twice.
            right = stack[-1]
            stack = stack[:-1]
            left = stack[-1]
            stack = stack[:-1]
            # Push operator applied to stack elements.
            stack = stack + [i(left, right)]
            
    # Should only be one item on stack.
    return stack[0]

In [28]:
import itertools as it

# Operators as functions.
import operator

# Example operators
operators = [operator.add, operator.mul, operator.sub, operator.add, operator.add]

# Using eval, which mightn't be great
list(map(eval_rpn, patterns(play_num, operators)))

[56,
 56,
 72,
 96,
 96,
 126,
 126,
 100,
 144,
 76,
 76,
 108,
 140,
 140,
 389,
 389,
 393,
 399,
 399,
 213,
 213,
 205,
 205,
 11,
 411,
 395,
 199,
 215,
 -292,
 -292,
 108,
 508,
 508,
 680,
 680,
 478,
 518,
 284,
 284,
 308,
 320,
 320]

## Explanation of the functional aspects of the code
***

***